Implement Partially Blind RSA 02

thibmeu opened the pull request:
Provide an implementation of [Partially Blind RSA draft 02](https://datatracker.ietf.org/doc/html/draft-amjad-cfrg-partially-blind-rsa-02). Note that while a lot of code could be shared between Blind RSA and Partially Blind RSA, we don't do this to have both implementation close to their respective specification.

The implementation is tested using draft test vectors.

During implementation, a few improvements have been identified for the draft, currently under review https://github.com/chris-wood/draft-amjad-cfrg-partially-blind-rsa/pull/49

armfazh commented:
interestingly, Node provides a method for [(safe) prime generation.](https://nodejs.org/api/crypto.html#cryptogenerateprimesize-options-callback)

thibmeu commented:
I've added the example, along with prime generation implementation.
I'm unsure yet how to update the README to mention this library provides implementation of two specs.
Performance optimisation might be part of a follow up ticket, especially given the spec is under work, and I'd like to develop Privacy Pass with extensions

armfazh commented on test/partially-blindrsa.test.ts:
```diff
@@ -0,0 +1,183 @@
+import { PartiallyBlindRSA, RSAPBSSA, getSuiteByName } from '../src/index.js';
+
+// Test vectors
+// https://www.rfc-editor.org/rfc/rfc9474.html#name-test-vectors
```
update link

armfazh commented on src/prime.ts:
```diff
@@ -0,0 +1,99 @@
+        bitLength = bitLength - 1;
+    }
+
+    const min = 2n ** BigInt(bitLength - 1);
```
here you can generate a random number by sampling random bytes with webcrypto
see: A.5 Converting Random Bits into a Random Number of https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf

so you can remove the `randomBigIntBetween` function

armfazh commented on test/partially-blindrsa.test.ts:
```diff
@@ -0,0 +1,183 @@
+
+    const params = [[], [{ supportsRSARAW: true }]];
+
+    test.each(params)(
+        `${v.name}`,
+        async (...params) => {
```
```suggestion
    describe.each(params)(`${v.name}`, (params) => {
        test(
            `supportsRSARAW/${params.supportsRSARAW}`,
            async () => {
```

armfazh commented on test/partially-blindrsa.test.ts:
```diff
@@ -0,0 +1,183 @@
+    });
+});
+
+describe.each(vectors)('TestVectors', (v: Vector) => {
```
```suggestion
describe.each(vectors)('TestVector_$#', (v: Vector) => {
```

thibmeu commented on src/partially-blindrsa.ts:
```diff
@@ -0,0 +1,450 @@
+            },
+            { ...algorithm, name: PartiallyBlindRSA.NAME },
+            true,
+            ['sign'],
```
adding verify prevents importing the private key `invalid key usages`. Therefore only using `sign`

thibmeu commented on src/index.ts:
```diff
@@ -49,6 +50,30 @@ export const Params: Record<string, BlindRSAParams> = {
         saltLength: 0,
         prepareType: PrepareType.Deterministic,
     },
+    RSAPBSSA_SHA384_PSS_Randomized: {
```
added

thibmeu commented:
I've addressed the comment from @armfazh review, as well as added a README.
The example sometimes passes, sometimes doesn't. The issue is 'The given number is not invertible.'. I wonder if this issue is in `inverseMod` implementation for odd numbers, or if the generated safe prime somehow do not work well. TBD.

thibmeu commented:
@armfazh I've merged your changes regarding field arithmetic logic for key generation. I've also rebased the changes onto `main`.

armfazh commented on src/partially_blindrsa.ts:
```diff
@@ -0,0 +1,489 @@
+                },
+                true,
+            );
+            return rsaRawBlingSign(privateKey, blindMsg);
+        }
+
+        // 3. s = RSASP1(sk_derived, m)
+        const s = rsasp1(sk_derived, m);
+
+        // 4. m' = RSAVP1(pk_derived, s)
+        const mp = rsavp1(pk_derived, s);
+
+        // 5. If m != m', raise "signing failure" and stop
+        if (m.equals(mp) === false) {
+            throw new Error('signing failure');
+        }
+
+        // 6. blind_sig = int_to_bytes(s, kLen)
+        // 7. output blind_sig
+        return i2osp(s, kLen);
```
here the function is returning the signature, and skipped the verification.
Note that in the specification, the signature is always checked, just in case the generation had produced an error.

thibmeu replied:
I'm not sure I totally get your point. The spec does not seem to mandate checking https://datatracker.ietf.org/doc/html/draft-amjad-cfrg-partially-blind-rsa-02#name-blindsign

The only thing that differs is that if the bitlength of the signature is bigger than the modulus length, int_to_bytes will error with a message different from the one defined by the spec.

armfazh replied:
the question here is whether `rsaRawBlingSign` processes these steps:

```
4. m' = RSAVP1(pk_derived, s)
5. If m != m', raise "signing failure" and stop
```

thibmeu replied:
reading the [platform code](https://github.com/cloudflare/workerd/blob/main/src/workerd/api/crypto-impl-asymmetric.c%2B%2B#L838-L879), I'm tempted to say no

we'll need to update [blind rsa](https://www.rfc-editor.org/rfc/rfc9474#name-blindsign) implementation as well, given the same pattern is used.

armfazh commented on src/partially_blindrsa.ts:
```diff
@@ -0,0 +1,489 @@
+        );
+
+        // 6. d = inverse_mod(e, phi)
+        // phi = (p-1)(q-1) = pq - p - q + 1 = n - p - q + 1
+        const d = inverseMod(e, phi);
```
```suggestion
        // phi = (p-1)(q-1) = pq - p - q + 1 = n - p - q + 1
        // TODO: replace this applying Chinese Remainder Theorem.
        const d = inverseMod(e, phi);
```

thibmeu replied:
comment added.

armfazh commented on src/partially_blindrsa.ts:
```diff
@@ -0,0 +1,489 @@
+        // 3. lambda_len = modulus_len / 2
+        // 4. hkdf_len = lambda_len + 16
+        const hkdf_input = joinAll([new TextEncoder().encode('key'), info, new Uint8Array([0x00])]);
+        const hkdf_salt = i2osp(n, n.bitLength() / 8);
+        const lambda_len = n.bitLength() / 8 / 2;
```
protect these two divisions with `Math.ceil`

thibmeu replied:
moved to bitshift

armfazh commented on src/partially_blindrsa.ts:
```diff
@@ -0,0 +1,489 @@
+        expanded_bytes[0] &= 0x3f;
+        expanded_bytes[lambda_len - 1] |= 0x01;
+
+        // 8. e' = bytes_to_int(slice(expanded_bytes, lambda_len))
```
this may be an error in the spec, because `slice` receives 3 parameters.
also, it's not clear why it requests additionally +16 bytes to the HDKF and they are not used.

thibmeu replied:
I agree. I've put a PR on the spec to correct this https://github.com/chris-wood/draft-amjad-cfrg-partially-blind-rsa/pull/49. This comment is a direct copy paste of draft 2 which is the latest spec atm

thibmeu replied:
I've updated it to `0, `, which is the same comment I made on the PR to the draft.

armfazh commented on src/partially_blindrsa.ts:
```diff
@@ -0,0 +1,489 @@
+        const phi = new sjcl.bn(sk.p).sub(1).mul(new sjcl.bn(sk.q).sub(1));
+
+        // 1. (n, e') = DerivePublicKey(n, info)
+        const pk_derived = await this.derivePublicKey({ n: sk.n, e: new sjcl.bn(0) }, info);
```
better to pass only `n`, note the spec is wrong passing both `n` and `e`, which is unknown.

thibmeu replied:
the spec mandates two parameters. I've suggested a change in https://github.com/chris-wood/draft-amjad-cfrg-partially-blind-rsa/pull/49

thibmeu replied:
I've updated the definition of derivePublicKey to be `Pick<BigPublicKey, "n">`

armfazh commented on src/partially_blindrsa.ts:
```diff
@@ -0,0 +1,489 @@
+        // 1. (n, e') = DerivePublicKey(n, info)
+        const pk_derived = await this.derivePublicKey({ n: sk.n, e: new sjcl.bn(0) }, info);
+
+        // 2. d' = inverse_mod(e', phi)
+        const d_prime = inverseMod(pk_derived.e, phi);
```
```suggestion
        // 2. d' = inverse_mod(e', phi)
        // Replace this with Chinese Remainder Theorem.
        const d_prime = inverseMod(pk_derived.e, phi);
```

armfazh commented on README.md:
```diff
@@ -2,9 +2,9 @@
 [![NPM](https://nodei.co/npm/@cloudflare/blindrsa-ts.png)](https://www.npmjs.com/package/@cloudflare/blindrsa-ts)
 
-# blindrsa-ts: A TypeScript Library for the RSA Blind Signature Protocol
+# blindrsa-ts: A TypeScript Library for the RSA (Plain and Partially) Blind Signatures Protocol
```
```suggestion
# blindrsa-ts: A TypeScript Library for Blind and Partially-Blind RSA Signature Protocols
```

armfazh reviewed (changes requested):
minor nits

thibmeu commented on src/util.ts:
```diff
@@ -275,3 +279,59 @@ export function random_integer_uniform(n: sjcl.bn, kLen: number): sjcl.bn {
 }
+
+// implement inverseMod for sjcl.bn where p is even
+export function inverseMod(x: sjcl.bn, p: sjcl.bn): sjcl.bn {
```
i've added a comment with the source https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm#Pseudocode

thibmeu commented:
@armfazh changes since your last review:
* address comments
* allow users to pass a method to generate safe prime. this is used in the example for instance by using node:crypto `generatePrimeSync({safe: true})`
* update test vectors naming following [merged of comments in the draft](https://github.com/chris-wood/draft-amjad-cfrg-partially-blind-rsa/pull/49)

armfazh commented on examples/partially_blindrsa.ts:
```diff
@@ -0,0 +1,135 @@
+    };
+    switch (prepare) {
+        case PrepareType.PreGenerated:
+            return preGeneratedKeys(true);
```
it looks like the function always passes `extractable=true`, better to remove the parameter.

armfazh reviewed (changes requested):
a few minor changes and it's ready

thibmeu commented on src/partially_blindrsa.ts:
```diff
@@ -0,0 +1,500 @@
+
+        // 2. sk_derived, pk_derived = DeriveKeyPair(sk, info)
+        const { secretKey: sk_derived, publicKey: pk_derived } = await this.deriveKeyPair(sk, info);
+
+        if (this.params.supportsRSARAW) {
+            const { privateKey } = await PartiallyBlindRSA.bigKeyPairToCryptoKeyPair(
+                { secretKey: sk_derived, publicKey: pk_derived },
+                {
+                    modulusLength: kLen * 8,
+                    publicExponent: new Uint8Array([1, 0, 1]),
+                    hash: this.params.hash,
+                },
+                true,
+            );
+            return rsaRawBlingSign(privateKey, blindMsg);
+        }
+
+        // 3. s = RSASP1(sk_derived, m)
+        const s = rsasp1(sk_derived, m);
+
```
I've done this both for Partially blind rsa and blind rsa (which presented the same behaviour). I've updated rsaRawBlingSign to return a sjcl.bn instead of a Uint8Array, for direct consumption by the algorithm.

thibmeu commented:
@armfazh thanks for the comments. I've addressed them all, and should be good for another pass.
